Add image color quantization and palette support

Xrayez opened the pull request:
Implemented as part of [ImageTools](https://github.com/Xrayez/godot-imagetools) module via `ImageIndexed` class

<details>
<summary>Initial proposal</summary>

![image-palettes](https://user-images.githubusercontent.com/17108460/56093229-8807d780-5ece-11e9-9cba-aa4853f92169.png)

Closes #27756.
Fixes #28020.
Related: #6532.


## Summary

This PR brings in an ability to create and generate an optimal color palette for an image and basic palette operations, allowing for things like finding dominant/average colors, palette swapping with/without using shaders etc. It also unifies and simplifies different image loaders and savers to use the same palette API, extract existing palette from images, and save PNG indexed images with transparency. This does not add `FORMAT_INDEXED` image format but rather complements the existing formats alongside. Some issues was revealed like `stex` format not supporting palette data that would store existing palette from imported images and wouldn't have to be loaded from within a script manually, which would make this feature more usable for pixel-art developers.

## Features

- [x] generate an optimal palette from an image:
  - with specified number of colors (max: 256);
  - optional dithering (ordered, random).
- [x] create palette and modify palette manually;
- [x] extract color palette from PNG images when loading from disk (instead of expanding);
- [x] save PNG with a palette associated with it:
  - the palette can be modified within Godot;
  - also supports saving indexed color images with "cheap" alpha.
- [x] extract existing palette from other image formats:
  - [x] bmp (8/4/1 bits)
  - [x] tga (8/4 bits)
- [x] implement palette formats:
    - [x] 32 bit
    - [x] 24 bit
- ~~make texture importer (`.stex`) to possibly embed palette data.~~
  - likely should not be part of this PR at least, see "Issues" section.
- ~~generate palette ordered (currently random):~~
  -  it's not ordered to distribution, might be a limitation of underlying quantization lib used...
  - requres modification of thirdparty lib...
  - if ordered, fetching least used colors would be possible, as well as easier to find dominant colors.

## Implementation details
This does **not** actually add a new `FORMAT_INDEXED` image format as modern GPUs don't actually support it as far as I know, but index data is stored internally alongside regular pixels *if* an image has a palette. This allows to interchange (swap, extend) index data to pixel data via `apply_palette()` method. Note that I've also purposely avoided creating and exposing methods for operating on individual indexed pixels for similar reasons (if such a need occurs, it's just better to do this outside of Godot in image editing software).

I've added palette load/save support for PNG and load for TGA/BMP formats. I believe different image loaders/savers could make use of the new palette API that way, [see the full list](https://en.wikipedia.org/wiki/Indexed_color#Image_file_formats_supporting_indexed_color). Note that all support 8-bit palette, so this should cover all cases, the palette bit depth is determined by format pixel size anyway.

I created it in a way that doesn't interfere with the existing image API, and in possibility that this feature is useless for most users so that I can branch off and write it in a separate module instead (not sure how easy would that be, so far it nicely integrates with existing codebase).

### Thirdparty library:
This feature uses [exoquant](https://github.com/exoticorn/exoquant) library for image quantization which is under MIT license and is very lightweight.

#### List of diffs
- removed `#include "malloc.h" as it's not available in iOS and MacOS.

## Issues

When image is imported, the original palette data (loaded from disk) is lost, as `stex` (custom Godot texture format) doesn't store any palette/index data, so one would only be able to retrieve existing palette from an image via `image.load()` or `image.load_png_from_buffer()`.

Here's my attempt to resolve this issue (in a sub-branch):
https://github.com/Xrayez/godot/commit/21f86cd87a73e81bc98593a0632aebc0c44329ca

If the issues in the above example could be resolved, another alternative would be also adding `num_colors` as import option and generate palette on import:
```
num_colors = -1 # default, indicate true color, or MAX_INT, INF...
if num_colors > 0:
    image.generate_palette(num_colors)
    image.apply_palette()
```

This method conveniently helps to implement #18662, which would make Godot to import textures with reduced number of colors automatically while still remaining true color internally, without using shaders at all as textures can always be reimported.


## Use cases and examples

* Taking a screenshot and reducing number of colors to optimize for size:
```gdscript
func save_screenshot():
	get_viewport().render_target_v_flip = true # hmm
	yield(VisualServer, "frame_post_draw")
	get_viewport().render_target_v_flip = false

	screenshot.convert(Image.FORMAT_RGBA8)
	screenshot.generate_palette()
	screenshot.save_png("screenshot_indexed.png")
```

* Finding dominant/average colors in an image:
```gdscript
image.generate_palette(2) # first color is usually background
var dominant_color = image.get_palette_color(1)
# or...
image.generate_palette(8)
var average_colors = image.get_palette()
```
![average-colors](https://user-images.githubusercontent.com/17108460/56416918-01227880-629b-11e9-9d7c-cbfa27b7adf3.png)


* Image posterization? (8-bit feel):
```gdscript
var image = get_node('sprite').texture.get_data()
image.generate_palette(8)
image.apply_palette()
```
![posterization](https://user-images.githubusercontent.com/17108460/56417674-48aa0400-629d-11e9-9edf-79e56d1c6ade.png)


*  (Pseudo) [Palette swapping](https://github.com/HeartoLazor/godot_palette_swap) #12670 (interchangeable with shaders):
```gdscript
var image = get_node('sprite').texture.get_data()
if not image.has_palette():
     image.generate_palette(16)

image.set_palette_color(5, Color.red)
image.set_palette_color(9, Color.gold)
image.set_palette_color(1, Color.darkred)

image.apply_palette()
```
![palette-swapping](https://user-images.githubusercontent.com/17108460/56417395-793d6e00-629c-11e9-91df-a122228825a9.png)


A color palette (array) could be passed to shaders as uniform, yet #10751 needs to be implemented to take advantage of this.

## Specific use cases:
* Suitable as a component for basic image editing in Godot.
* Creating user generated content that must be cross/backward-compatible (can't deal with transparency in images and/or require limited number of colors).

## Example project

You can try/test these features out here:
https://github.com/Xrayez/godot-image-indexed-example

</details>

Commit:
Initial image quantizer implementation

Commit:
Add indexed format, save/load images with palette (png)

Commit:
FORMAT_INDEX attempts and png import/export with palette improvements

Commit:
Make palette as raw data

Commit:
Fix get_palette, map image to indexed data

Commit:
Add palette to rgba method

Commit:
Properly read PNG with new palette API

Commit:
Apply palette to image without making copy, add sanity checks

Commit:
Fix get_palette, make ImageSaverPNG work again

Commit:
Avoid possible loss of precision when loading png palette

Set palette data directly instead.

Commit:
Do the same treatmeant for image saver

Commit:
Fix set/get palette color using wrong offset

Commit:
Expose dithering modes

Commit:
Clamp number of colors to [1,256] range, add param for specifying alpha

Commit:
More clear assertion on attempting to generate palette from an empty image

Commit:
Return mean error from generated palette

Commit:
Init quantizer inside method, don't be defensive when freeing png palette

Commit:
Unify palette and index data creation, ensure palette pixel size is always 4 bytes

Calinou commented:
This looks great!

> Finding dominant/average colors in an image:

That can be pretty useful actually, even outside of pixel art games :slightly_smiling_face:

Commit:
Attempt to resolve compilation on iOS and Mac

https://stackoverflow.com/questions/5929787/include-malloc-h-xcode

fire commented:
How different is this from HDR to LDR? Is there anything we can reuse from here?

Xrayez commented:
@fire I don't think there's connection, what you mentioned involves more sophisticated image processing rather than simple color quantization (to my limited knowledge). `generate_palette` is another name for `quantize` method, which just clusters similar colors, hence the need to create color palette, which naturally leads to adding pseudo-support for indexed images (pseudo because images are still true color internally).

This is more for 2D rather than 3D if you were to ask.

Xrayez commented:
All in all, I think this can help implement image importers better, as some "drivers" might not have built-in support for expanding indexed images to true color out-of-the-box (#28020, #17261).

Commit:
Import 8-bit bmp images

- added some sanity checks according to bmp specification;

- properly find color table offset, don't depend on sizeof(struct) as it's implementation defined related to structure padding;

- read color table and index data within the same scope, and then simply extend palette using the new palette API.

Commit:
Import 1-bit and 4-bit bmp images

- not all images can be imported as it requires bit unpacking (size dimensions must be a multiple of 8 for 1-bit and 2 (even) for 4-bit images).

Commit:
Fix -Wimplicit-fallthrough CI warnings in bmp loader

Xrayez commented:
Well I managed to properly extract palette colors from 8/4/1-bit bmp images while getting benefits from `apply_palette` to expand these indexed images easily, which proves my statement about how different bit depth can be converged to the same base.

Looking through other image loaders that support indexed images, I've stumbled upon TGA, and noticed this:
https://github.com/godotengine/godot/blob/b16c309f82c77d606472c3c721a1857e323a09e7/modules/tga/image_loader_tga.cpp#L150-L156

The developer has to manually remap palette with indices to extend it which could be easily done with mentioned `apply_palette` method! So this would avoid duplicate code in image loaders.

Commit:
Extract color palette from TGA indexed images

NOTE: this commit fixes incorrect color table lookup where red and blue channels were flipped.

Commit:
Fix PNG loader inconsistencies loading indexed images with and without alpha chunk

Commit:
More complete support for 24/32 bit depth palette manipulation

Palette bit depth is determined by image format pixel size instead which should cover more potential formats (if such support is added in the future).

Commit:
Fix saving PNG indexed images with RGB8 format

Two-Tone commented:
What's transparency support like? I assume it's 1 bit like everywhere else?

Xrayez commented:
@Two-Tone according to my research of indexed formats, most don't support transparency component in a color palette/table/map, but PNG allows to use tRNS chunk (alpha, 1 byte)  along with PLTE chunk (palette, 3 bytes) to add transparency to indexed images, which most image decoders will interpret as "true alpha", so I do make use of it when saving RGBA images with associated palette via `save_png`. Here I've rewritten ImageLoader/SaverPNG to read and write those chunks directly, instead of expanding palette to color on reading, as well as BMP/TGA but those don't support indexed transparency.

Xrayez commented:
To demonstrate, here's a Godot logo loaded as true color and saved as indexed PNG with quantized 32 colors and transparency (all from within Godot):

![logo](https://user-images.githubusercontent.com/17108460/56349065-7e34eb80-61d0-11e9-9828-32574c50d48f.png)

If you open it in GIMP, you'll see that it has transparency, yet the color palette will only have those colors. The size is also reduced from 12 Kb to 5 Kb.

Commit:
More palette and index data validation and fixes

- introduce MAX_PALETTE_SIZE constant (256);
- don't allow to set palette manually with different size (palette must be present);
- fix offsets when setting palette per format;
- invalidate palette by clearing it when converting between image formats.

Commit:
Generate mipmaps (if used) on applying palette

Two-Tone commented:
That's great, although opening it in gimp just results in 1 bit transparency, for future reference.

Xrayez commented:
I've done some attempts on embedding palette and index data to `stex` textures in a separate branch:

https://github.com/Xrayez/godot/commit/21f86cd87a73e81bc98593a0632aebc0c44329ca

Supporting this would be useful for pixel-art developers. An indexed image could be created in image editing software outside of Godot, imported automatically and directly into `stex` texture, and retrieved with `texture.get_data().get_palette()`, from there many palette operations could be performed on such image, but the current implementation indicates that there are a lot to consider, so I doubt this should be supported at all...

I'm done with implementing features, so suspending my work until I get feedback from maintainers!

Test project to save your day:
https://github.com/Xrayez/godot-image-paletted-example

reduz commented:
No, I really don't want this. I removed palette support in Godot 3.0 because the actual use cases for this are near null. There are better ways to do these effects in modern games using masking. If you want something like this, I suggest writing a GDNative add-on or similar.

Xrayez commented:
Understandable... I was going to write this in a module anyway, but making first attempts was logical to do within Image class itself.

Looking at how other image loaders are implemented, I think I could hopefully override PNG specific callbacks in order to extract and save indexed PNG images with modified palette... Maybe some Image methods could be made virtual, will see.

I'm going to extract some bug fixes I've discovered during developing this PR then 😛

Two-Tone commented:
> No, I really don't want this. I removed palette support in Godot 3.0 because the actual use cases for this are near null. 

What about #18269? The ability to simply import GIF animations and do palette swapping is something a lot of the MUGEN users tout as being very, very useful. Sure, you can do palette swapping via shaders, but that is incredibly cumbersome and there would still be no way of importing GIFs.

Xrayez commented:
Implemented as part of [ImageTools](https://github.com/Xrayez/godot-imagetools) module via `ImageIndexed` class. Also added more methods to manipulate indexed data such as `set_pixel_indexed` and `get_pixel_indexed` that works similarly to `set_pixel` and `get_pixel`, and in fact both can be used in extended `ImageIndexed` class as index data and image data are stored alongside and can be swapped.

I've tried to write this in GDNative but I've stumbled upon some limitations/inconveniences that didn't allow me to reuse the engine internals to full extent, yet I managed to extract image color quantization routines quite easily which GDNative seems like a good fit, namely exposing methods to generate color palette from image for finding average/dominant colors efficiently, which some noted (me included) to be quite useful. Still learning and working on setting up a proper build infrastructure for GDNative though. Cheers!

CarlGustavAlbertDwarfsteinYung commented:
I use palettes in my game. I wish I had some of this functionality but ah well. I am implementing it myself in gdscript instead.

aaronfranke commented:
What exactly prevents this functionality from being in a GDScript or GDNative plugin?

> allow me to reuse the engine internals to full extent

Have you considered proposing modifications to GDNative to expose more of the API?

Xrayez commented:
> What exactly prevents this functionality from being in a GDScript or GDNative plugin?

Nothing. Same goes for some other classes and built-in nodes, see [godot-gdscript-ports](https://github.com/godot-extended-libraries/godot-gdscript-ports) for example.

> Have you considered proposing modifications to GDNative to expose more of the API?

My reasoning is the same as in godotengine/godot-proposals#565. I'd rather use C++ modules, and even proposed a ready solution to help this #36922, but I'm not sure if it's going to be merged because https://github.com/godotengine/godot-proposals/issues/565#issuecomment-596133231:

> For the rest, I really think it becomes a much more corner use cases, which does not make sense for having an entire feature for, given the idea is to discourage the use of modules as much as possible towards the future.

---

> it doesn't seem like a feature that everyone would use

That's my problem, I find it difficult to tell which features would be used frequently. I definitely use those features frequently myself (not to say that's a good reason to deem a feature as frequent, but as example of vagueness of this requirement).

Hence the need for godotengine/godot-proposals#575, I fully expose my ignorance and ask core devs to make this process (design philosophy, the direction of the project, what goes to core vs plugin etc) as clear as possible, so the teachings is not on the level of GitHub comments/IRC only, but on the level of documentation. 🙂

dawdle-deer commented:
Can we revisit this? I'll lay out a concrete use case here, if it helps:

I'm developing VisualNML, a visual scripting language for modding OpenTTD (an open-source reimplementation of Transport Tycoon Deluxe, a DOS game from the 90s with a surprisingly-vibrant and active modding community, even decades later).

At the time, technical limitations constrained sprites to an 8bpp indexed format. However, even though the game now supports 16- and 32-bit color images, the majority of assets being created for mods still use the original 8-bit color palettes, as this technical limitation has become an aesthetic style and creative challenge.

A core feature of VNML is an integrated painting app that leans into an indexed sprite palette by default. This is useful over external apps like GIMP because certain color indices are set to render with hard-coded special effects (including animated water, neon light patterns, and palette-substitution for player color), and with some custom shader effects, the paint editor benefits from live previews of these effects, _exactly_ as they will appear in-game.

I've worked around several minor points of friction up to this point, but the big hurdle I'm running up against now is PNG export. As it stands, Godot does not have a friendly way to save a PNG file that uses a color palette. Goost seems fine, but `PaletteImage` is not a drop-in replacement for my project, and methods like [ResourceSaverIndexedPNG::save_image](https://github.com/goostengine/goost/blob/gd3/core/image/drivers/png/resource_saver_indexed_png.cpp) aren't easy to detangle from the rest of Goost for my own purposes. Moreover, I don't want to introduce Goost as a dependency for contributing to VNML. And, the other workarounds I have available to me don't seem much easier (bundling the project with a converter app and executing it, writing my own driver extension and including it).

I know this is _quite_ the necropost. I'm also expecting that revitalizing this would warrant a complete reimplementation. But even just getting extra parameters in `save_png` to support writing a palette into a png file would go a _very_ long way to help me finish this project. I'd even be happy to handle implementation, so long as it gets merged eventually (it's a non-starter to make this project depend on a fork of the engine that I need to personally maintain in perpetuity).
